dbms/src/Storages/tests/gtest_SplitTokenExtractor.cpp

```diff
@@ -0,0 +1,174 @@
+#include <Storages/MergeTree/MergeTreeIndexFullText.h>
+
+#include <Common/PODArray_fwd.h>
+#include <Common/PODArray.h>
+
+#include <gtest/gtest.h>
+
+#include <vector>
+#include <string_view>
+#include <string>
+#include <memory>
+
+namespace
+{
+using namespace DB;
+}
+
+struct SplitTokenExtractorTestCase
+{
+    const char * description;
+    const std::string source;
+    const std::vector<std::string> tokens;
+};
+
+std::ostream & operator<<(std::ostream & ostr, const SplitTokenExtractorTestCase & test_case)
+{
+    return ostr << test_case.description;
+}
+
+class SplitTokenExtractorTest : public ::testing::TestWithParam<SplitTokenExtractorTestCase>
+{
+public:
+    void SetUp() override
+    {
+        const auto & param = GetParam();
+        const auto & source = param.source;
+        data = std::make_unique<PaddedPODArray<const char>>(source.data(), source.data() + source.size());
+    }
+
+    std::unique_ptr<PaddedPODArray<const char>> data;
+};
+
+TEST_P(SplitTokenExtractorTest, next)
+{
+    const auto & param = GetParam();
+
+    SplitTokenExtractor token_extractor;
+
+    size_t i = 0;
+
+    size_t pos = 0;
+    size_t token_start = 0;
+    size_t token_len = 0;
+
+    for (const auto & expected_token : param.tokens)
+    {
+        SCOPED_TRACE(++i);
+        ASSERT_TRUE(token_extractor.next(data->data(), data->size(), &pos, &token_start, &token_len));
+        EXPECT_EQ(expected_token, param.source.substr(token_start, token_len))
+                << " token_start:" << token_start << " token_len: " << token_len;
+    }
+
+    ASSERT_FALSE(token_extractor.next(data->data(), data->size(), &pos, &token_start, &token_len));
+}
+
+// Helper to allow strings with embedded '\0' chars.
+#define BINARY_STRING(str) std::string{str, sizeof(str) - 1}
```

Comment:
You can use c++17 suffix literal feature

https://en.cppreference.com/w/cpp/string/basic_string/operator%22%22s

https://gcc.godbolt.org/z/HY3nwV

Reply:
Cool feature, but again, this is a C++20...

Reply:
Nope, this is a C++14 feature with C++20 improvements. Also in Clickhouse there is alread C++20

Reply:
Okay, fixed